subsys/bluetooth/host/l2cap.c

```diff
@@ -3065,6 +3072,16 @@ static int bt_l2cap_dyn_chan_send(struct bt_l2cap_le_chan *le_chan, struct net_b
 		return -EMSGSIZE;
 	}
 
+	if (buf->ref != 1) {
+		/* The host may alter the buf contents when segmenting. Higher
+		 * layers cannot expect the buf contents to stay intact. Extra
+		 * refs suggests a silent data corruption would occur if not for
+		 * this error.
+		 */
+		LOG_ERR("buf given to l2cap has other refs");
+		return -EINVAL;
+	}
```

Comment:
Would it make sense to move this to a function that is used by both this and the conn_cb function? 

Just to avoid duplicating the check and comment

Reply:
Not sure if this makes sense. Were you thinking of a function hiding the `buf->ref != 1` with a semantic name? `buf_unsafe_to_clobber`? Did you have a specific name in mind?

The comments are slightly different btw. One pertains to segmenting and the other to fragmenting.

Reply:
> Did you have a specific name in mind?

Thinking that it could be a function in e.g. `buf.c` with something along the name of `bt_buf_can_send` or `bt_buf_valid_tx`. In the case we want to add additional checks that are common we can add it to a single function, rather than to multiple. 

Not sure if valid, but other checks that could be collected there are things like `user_data_size == CONFIG_BT_CONN_TX_USER_DATA_SIZE`, `len <= size`, `data != NULL`, etc.